oak_runtime/src/node/http/user.rs

```diff
@@ -0,0 +1,303 @@
+                .handle,
+        )?;
+
+        // Close all local handles except for the one that allows reading responses.
+        pipe.close(&runtime);
+
+        Ok(pipe.response_reader)
+    }
+
+    fn try_run(
+        self: Box<Self>,
+        runtime: RuntimeProxy,
+        startup_handle: oak_abi::Handle,
+    ) -> Result<(), ()> {
+        let receiver = Receiver::<OuterHttpInvocation>::new(ReadHandle {
+            handle: startup_handle,
+        });
+        let outer_http_invocation = receiver.receive(&runtime).map_err(|err| {
+            error!(
+                "{}: Failed to retrieve the outer HTTP invocation: {:?}",
+                self.node_name, err
+            )
+        })?;
+
+        // Send the request to the Oak node and get a handle to the channel containing the response.
+        let response_reader =
+            self.inject_http_request(runtime.clone(), outer_http_invocation.clone())?;
+        let response_receiver = crate::io::Receiver::new(ReadHandle {
+            handle: response_reader,
+        });
+        let response = response_receiver.receive(&runtime).map_err(|err| {
+            error!(
+                "Could not receive the response from the Oak node on channel {}: {}.",
+                response_reader, err
+            )
+        })?;
+
+        outer_http_invocation
+            .response_sender
+            .ok_or(())
+            .and_then(|sender| {
+                sender.send(response, &runtime).map_err(|err| {
+                    warn!(
+                        "{}: Could not send response to HTTP server pseudo-node: {}",
+                        self.node_name, err,
+                    )
+                })
+            })
+    }
+}
+
+impl Node for UserNode {
+    fn node_type(&self) -> &'static str {
+        "user-node"
+    }
+
+    fn run(
+        self: Box<Self>,
+        runtime: RuntimeProxy,
+        startup_handle: oak_abi::Handle,
+        _notify_receiver: oneshot::Receiver<()>,
+    ) {
+        let _unit_result = self.try_run(runtime, startup_handle);
+    }
+
+    fn get_privilege(&self) -> NodePrivilege {
+        self.privilege.clone()
+    }
+}
+
+/// A pair of temporary channels to pass the HTTP request to the Oak node, and to receive the
+/// response.
+struct Pipe {
+    request_writer: oak_abi::Handle,
+    request_reader: oak_abi::Handle,
+    response_writer: oak_abi::Handle,
+    response_reader: oak_abi::Handle,
+}
+
+impl Pipe {
+    fn new(
+        runtime: &RuntimeProxy,
+        request_label: &Label,
+        privilege: &NodePrivilege,
+    ) -> Result<Self, ()> {
+        let request_reader_label = Label {
+            confidentiality_tags: request_label.confidentiality_tags.clone(),
+            integrity_tags: privilege
+                .can_endorse_integrity_tags
+                .clone()
+                .into_iter()
+                .collect(),
+        };
+        let response_writer_label = Label {
+            confidentiality_tags: privilege
```

Comment:
Should this check whether user identity tag was included in the original label? If the user did not include the identity in the original request, this could unnecessarily increase the confidentiality.